### ISSUE TITLE
Client of dynamic scale stream stalled after a while

### ISSUE THREAD

**Issue by vroy** [pravega/pravega#1806]
**Problem description**

We have an application that is running 3 writers and 3 readers in separate threads. Each writers write ~10 events per second, 100 bytes each.

After a while, around `2017-08-27 17:29:43`, each writers seemed to have blocked in `EventStreamWriterImpl::writeEventInternal`. After that, there are no more logs from the pravega client package.

I attached the logs that we had available. Unfortunately we don't have controller logs from around the time where the client appears to stall. I've also attached thread dumps in case it helps.

[client-writers-stall.zip](https://github.com/pravega/pravega/files/1257432/client-writers-stall.zip)

Another thing noticed in this environment is that even though there doesn't appear to be any more successful writes, the stream does not scale back down.

![image](https://user-images.githubusercontent.com/4655/29782658-cba441a4-8bf3-11e7-83bb-e523b492e2b9.png)

**Comment by vroy** [comment]
Potentially related: #1803 and #1792

**Comment by jkhalack** [comment]
The last scaling event reported by the scaling-events API is the following:
```
[{"timestamp":1503852902124,"segments":[{"number":2000,"start":1503851697230,"keyStart":0.7604775365479084,"keyEnd":0.7890304434280885},{"number":2001,"start":1503851697230,"keyStart":0.7890304434280885,"keyEnd":0.8175833503082686},{"number":2003,"start":1503851947539,"keyStart":0.5579178825361473,"keyEnd":0.7604775365479084},{"number":2004,"start":1503852079431,"keyStart":0.0,"keyEnd":0.09808164637255569},{"number":2008,"start":1503852369134,"keyStart":0.36048154824859846,"keyEnd":0.38640370556272036},{"number":2009,"start":1503852369134,"keyStart":0.38640370556272036,"keyEnd":0.41232586287684225},{"number":2010,"start":1503852717040,"keyStart":0.09808164637255569,"keyEnd":0.14712246955883354},{"number":2011,"start":1503852717040,"keyStart":0.14712246955883354,"keyEnd":0.19616329274511138},{"number":2012,"start":1503852759230,"keyStart":0.8175833503082686,"keyEnd":0.9087916751541343},{"number":2013,"start":1503852759230,"keyStart":0.9087916751541343,"keyEnd":1.0},{"number":2014,"start":1503852781536,"keyStart":0.41232586287684225,"keyEnd":0.48512187270649476},{"number":2015,"start":1503852781536,"keyStart":0.48512187270649476,"keyEnd":0.5579178825361473},{"number":2016,"start":1503852772828,"keyStart":0.19616329274511138,"keyEnd":0.2783224204968549},{"number":2017,"start":1503852772828,"keyStart":0.2783224204968549,"keyEnd":0.36048154824859846}]}]
```
The event timestamp is `1503852902124` (Aug 27 2017 16:55:02 GMT), so the last scaling event happened ~35 min before the last client log entry.


The two last segments (2016 and 2017) have the start time of `1503852772828` (Aug 27 2017 16:52:52 GMT, 2 minutes 20 seconds before the scaling event was registered).
Segments 2014 and 2015 have the start time of `1503852781536` (Aug 27 2017 16:53:01 GMT,  2 minutes 01 second before the scaling event timestamp).

**Comment by tkaitchuck** [comment]
Is something constantly sealing segments in this test? In the logs the latest segment numbers the writers are writing is over 1700.

**Comment by vroy** [comment]
The test is designed to scale up and down often. We don't explicitly seal segments as far as I can tell. This is the stream scaling policy that we apply and each writer writes ~10 events per second:

```
Scaling: Dynamic
Min Segments: 1
Scaling Trigger: Incoming Event Rate
Scaling Target Rate: 3 Events / second
Scaling Factor: 2
```

This is what it looks like when it's working:
![image](https://user-images.githubusercontent.com/4655/29849854-fbae789e-8cff-11e7-86a5-fbff7dabbe80.png)

**Comment by maddisondavid** [comment]
[Hulk-Pravega-Connection-Deadlock.txt](https://github.com/pravega/pravega/files/1267463/Hulk-Pravega-Connection-Deadlock.txt)

The attached log shows the client logs during one of these stalls.  As you can see, the number of events being generated by the client goes up until the client connection issue happens; after which the number of events being written stalls.

[threaddump.txt](https://github.com/pravega/pravega/files/1267466/threaddump.txt) Shows the threads at this time.  The writers all appear to be stuck attempting to get the `      writeFlushLock.writeLock().lock();`.  Since all other accessors of the lock are for the readLock we can assume a reader accessor has the lock.

[SegmentStore.txt](https://github.com/pravega/pravega/files/1267501/SegmentStore.txt)
handleLogSealed is one such place and in fact this can be seen in the ScalingRetransmission thread and in the SegmentStore.txt we can see the segmentstore around the same time saying the log is sealed:

```
"ScalingRetransmition-hulk/smallScale-1162" #5524 daemon prio=5 os_prio=0 tid=0x00007fcc80009800 nid=0x15ab waiting on condition [0x00007fcc6d0d5000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000ec3d3f48> (a java.util.concurrent.CompletableFuture$Signaller)
	at java.util.concurrent.locks.LockSupport.park(Unknown Source)
	at java.util.concurrent.CompletableFuture$Signaller.block(Unknown Source)
	at java.util.concurrent.ForkJoinPool.managedBlock(Unknown Source)
	at java.util.concurrent.CompletableFuture.waitingGet(Unknown Source)
	at java.util.concurrent.CompletableFuture.get(Unknown Source)
	at io.pravega.common.concurrent.FutureHelpers.getThrowingException(FutureHelpers.java:154)
	at io.pravega.client.segment.impl.SegmentOutputStreamImpl.write(SegmentOutputStreamImpl.java:376)
	- locked <0x00000000ec3d3ee8> (a java.lang.Object)
	at io.pravega.client.stream.impl.EventStreamWriterImpl.resend(EventStreamWriterImpl.java:191)
	at io.pravega.client.stream.impl.EventStreamWriterImpl.lambda$handleLogSealed$1(EventStreamWriterImpl.java:169)
	at io.pravega.client.stream.impl.EventStreamWriterImpl$$Lambda$155/7382024.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
```
The last thing this thread does before being blocked is `                connection = FutureHelpers.getThrowingException(getConnection());` which gets a future that completes when the connection is set back up.  Since this never completes something happens which means the connection future never gets completed and that stalls all the threads.

The following shows a stack trace from the client when the connection has problems (and from which the client never recovers)

```
2017-08-31 11:04:21,893 165402941 [pool-4-thread-2] WARN  i.p.c.s.i.SegmentOutputStreamImpl - Connection 96433b74-34a2-4c53-9681-9883dbfa5be1 failed due to:
io.pravega.shared.protocol.netty.ConnectionFailedException: java.nio.channels.ClosedChannelException
        at io.pravega.common.concurrent.FutureHelpers.getAndHandleExceptions(FutureHelpers.java:182)
        at io.pravega.client.netty.impl.ClientConnectionInboundHandler.send(ClientConnectionInboundHandler.java:114)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl.write(SegmentOutputStreamImpl.java:387)
        at io.pravega.client.stream.impl.EventStreamWriterImpl.writeEventInternal(EventStreamWriterImpl.java:126)
        at io.pravega.client.stream.impl.EventStreamWriterImpl.writeEvent(EventStreamWriterImpl.java:110)
        at com.emc.nautilus.pravega.taskexecutor.workers.writers.PravegaWriter.writeEvent(PravegaWriter.java:248)
        at com.emc.nautilus.pravega.taskexecutor.workers.writers.PravegaWriter.start(PravegaWriter.java:157)
        at com.emc.nautilus.pravega.taskexecutor.workers.writers.WriterWorkerManager.lambda$start$0(WriterWorkerManager.java:121)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: java.nio.channels.ClosedChannelException: null
        at io.netty.channel.AbstractChannel$AbstractUnsafe.close(...)(Unknown Source)
2017-08-31 11:04:21,893 165402941 [pool-4-thread-1] WARN  i.p.c.s.i.SegmentOutputStreamImpl - Connection 3fa3c2e7-4dbd-40c6-8aeb-4f0c0b429565 failed due to:
io.pravega.shared.protocol.netty.ConnectionFailedException: java.nio.channels.ClosedChannelException
        at io.pravega.common.concurrent.FutureHelpers.getAndHandleExceptions(FutureHelpers.java:182)
        at io.pravega.client.netty.impl.ClientConnectionInboundHandler.send(ClientConnectionInboundHandler.java:114)
        at io.pravega.client.segment.impl.SegmentOutputStreamImpl.write(SegmentOutputStreamImpl.java:387)
        at io.pravega.client.stream.impl.EventStreamWriterImpl.writeEventInternal(EventStreamWriterImpl.java:126)
        at io.pravega.client.stream.impl.EventStreamWriterImpl.writeEvent(EventStreamWriterImpl.java:110)
        at com.emc.nautilus.pravega.taskexecutor.workers.writers.PravegaWriter.writeEvent(PravegaWriter.java:248)
        at com.emc.nautilus.pravega.taskexecutor.workers.writers.PravegaWriter.start(PravegaWriter.java:157)
        at com.emc.nautilus.pravega.taskexecutor.workers.writers.WriterWorkerManager.lambda$start$0(WriterWorkerManager.java:121)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: java.nio.channels.ClosedChannelException: null
        at io.netty.channel.AbstractChannel$AbstractUnsafe.close(...)(Unknown Source)
```

**Comment by tkaitchuck** [comment]
Action items on this issue:
1. We should tone down error level in AppendProcessor and PravagaCommandProcessor for normal events. 
2. Odd exception for zero length read. https://github.com/pravega/pravega/issues/1832 @andreipaduroiu - can you check if this could be related.
3. Store.getStreamSegmentInfo with waitForPendingOps set to true does not result in a APPEND_SETUP being sent to the client or a connection failure. Looking into this, I'm fairly confident it isn't client side, because the replies are logged first thing when they come over the wire, before we even attempt to process them. @andreipaduroiu - Can you help me with this?
4. Client needs to time out connections as a failsafe. - I'll take this one.
5. Controller is stuck scaling, attempting to perfrom the same operations over and over for some reason. @shiveshr - Can  you look into this?

**Comment by chrisdail** [comment]
Linking potential fix #1865

**Comment by tkaitchuck** [comment]
Resolving as 1865 appears to have fixed it. Please open a new issue if this recurs.
